safecast_deploy/same_env.py

```diff
@@ -2,106 +2,125 @@
 import pprint
 import sys
 
-from safecast_deploy import git_logger, verbose_sleep
+from safecast_deploy import aws_state, verbose_sleep
+from safecast_deploy.aws_state import AwsTierType
+from safecast_deploy.exceptions import EnvNotHealthyException, EnvUpdateTimedOutException
 
 
 class SameEnv:
-    def __init__(self, state):
-        self.state = state
-        self._c = state.eb_client
+    def __init__(self, target_env_type, old_aws_state, new_aws_state, eb_client, result_logger,
+                 update_wait=70, total_update_wait=480):
+        self._target_env_type = target_env_type
+        self._aws_app_name = old_aws_state.aws_app_name
+        self._old_aws_env_state = old_aws_state.envs[target_env_type]
+        self._new_aws_env_state = new_aws_state.envs[target_env_type]
+        self._c = eb_client
+        self._result_logger = result_logger
+        self._update_wait = update_wait
+        self._total_update_wait = total_update_wait
 
     def run(self):
         self.start_time = datetime.datetime.now(datetime.timezone.utc)
-        # Handle the worker environment first, to ensure that database
+
+        self._check_environments()
+
+        # Handle the worker tier first, to ensure that database
         # migrations are applied
-        self._handle_worker()
-        self._handle_web()
-        result = self._generate_result()
-        self._print_result(result)
-        git_logger.log_result(result)
+        self._update_environment(AwsTierType.WORKER)
+        self._update_environment(AwsTierType.WEB)
 
-    def _handle_worker(self):
-        if self.state.has_worker:
-            print("Deploying to the worker.", file=sys.stderr)
-            env_name = self.state.env_metadata[self.state.subenvs['wrk']]['name']
-            self._update_environment(env_name)
+        result = self._generate_result()
+        self._result_logger.log_result(result)
 
-    def _handle_web(self):
-        print("Deploying to the web instances.", file=sys.stderr)
-        env_name = self.state.env_metadata[self.state.subenvs['web']]['name']
-        self._update_environment(env_name)
+    # Make sure we're not trying to deploy on top of an environment in distress
```

Comment:
I wonder about this comment. Could there be a case we need a deploy to get us out of a bad state?

Reply:
Yes, definitely.

Unfortunately, there's a lot of reasons that an environment could be in a failing state. I added this check because a worker environment was in a warning state due to messages in its dead-letter queue. Before I added this check, the new code would be deployed to the worker, but `safecast_deploy` would then spin for several minutes waiting for the environment to become green again before ultimately failing (since the new deploy didn't fix the problem). Sometimes it also seems like there's a significant lag between fixing a problem and seeing health status come back to green as well.

So in general, I think it's a good idea to fail quickly and make the operator track down what the problem is.

Right now, it's also possible to use `new_env` to deploy out to a new environment, bypassing this check... which could run into the same problems with things like dead-letter queues. That's a potential area for enhancement as well.

Finally, we could enhance the code to allow overriding this check (using a CLI flag) in a subsequent PR. Would also probably be a good idea to add a `restart` command to the deployer, although that's pretty easy through the Web UI as well.

I think these should all be taken as enhancements for the future.

Reply:
> Finally, we could enhance the code to allow overriding this check (using a CLI flag) in a subsequent PR.

this approach makes sense to me. 👍